src/devices/GoPiGo3/GoPiGo3.cs

```diff
@@ -0,0 +1,900 @@
+            byte[] reply = SpiTransferArray(outArray);
+            Console.WriteLine($"{reply[0]} {reply[1]} {reply[2]} {reply[3]} {reply[4]}");
+            return (reply[3] == 0xA5) ? reply[4] : throw new IOException($"{nameof(SpiRead16)} : no SPI response");
+        }
+
+        /// <summary>
+        /// Send a 8 bit value over SPI
+        /// </summary>
+        /// <param name="MessageType">The SPI message type</param>
+        /// <param name="Value">The value to be sent</param>
+        public void SpiWrite8(SpiMessageType MessageType, int Value)
+        {
+            byte[] outArray = { SpiAddress, (byte)MessageType, (byte)(Value & 0xFF) };
+            SpiTransferArray(outArray);
+        }
+
+        /// <summary>
+        /// Send a 16 bit value over SPI
+        /// </summary>
+        /// <param name="MessageType">The SPI message type</param>
+        /// <param name="Value">The value to be sent</param>
+        public void SpiWrite16(SpiMessageType MessageType, int Value)
+        {
+            byte[] outArray = { SpiAddress, (byte)MessageType, (byte)((Value >> 8) & 0xFF), (byte)(Value & 0xFF) };
+            SpiTransferArray(outArray);
+        }
+
+        /// <summary>
+        /// Send a 24 bit value over SPI
+        /// </summary>
+        /// <param name="MessageType">The SPI message type</param>
+        /// <param name="Value">The value to be sent</param>
+        public void SpiWrite24(SpiMessageType MessageType, int Value)
+        {
+            byte[] outArray = { SpiAddress, (byte)MessageType, (byte)((Value >> 16) & 0xFF), (byte)((Value >> 8) & 0xFF), (byte)(Value & 0xFF) };
+            SpiTransferArray(outArray);
+        }
+
+        /// <summary>
+        /// Send a 32 bit value over SPI
+        /// </summary>
+        /// <param name="MessageType">The SPI message type</param>
+        /// <param name="Value">The value to be sent</param>
+        public void SpiWrite32(SpiMessageType MessageType, int Value)
+        {
+            byte[] outArray = { SpiAddress, (byte)MessageType, (byte)((Value >> 24) & 0xFF), (byte)((Value >> 16) & 0xFF), (byte)((Value >> 8) & 0xFF), (byte)(Value & 0xFF) };
+            SpiTransferArray(outArray);
+        }
+
+        #endregion
+
+        #region Board elements
+
+        /// <summary>
+        /// Read the 20 characters of GoPiGo3 manufacturer name
+        /// </summary>
+        /// <returns>Returns the GoPiGo3 manufacturer name string</returns>
+        public string GetManufacturer()
+        {
+            string retVal = string.Empty;
+            byte[] outArray = {SpiAddress, (byte)SpiMessageType.GetManufacturer,
+                  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 };
+            byte[] reply = SpiTransferArray(outArray);
+            if (reply[3] != 0xA5)
+            {
+                throw new IOException("No SPI response");
+            }
+            else
+            {
+                for (int ndx = 4; ndx < 24; ++ndx)
```

Comment:
You can most likely use Encoding.UTF8/Encoding.ASCII GetString for this (not sure about 0 character but if it doesn't do length correctly I'd still search for 0 first and then call it). If not then StringBuilder should be much faster

Reply:
This is directly working:
```csharp
return Encoding.ASCII.GetString(reply.Skip(4).ToArray());
```
not sure it's much faster in terms of performance but it's shorter code

Reply:
@Ellerbach see: https://docs.microsoft.com/en-us/dotnet/api/system.text.encoding.getstring?view=netcore-2.1#System_Text_Encoding_GetString_System_Byte___System_Int32_System_Int32_